plutus-ledger/src/Ledger/Tx/CardanoAPI.hs

```diff
@@ -419,11 +421,19 @@ fromCardanoTxOut (C.TxOut addr value datumHash) =
     <*> pure (fromCardanoTxOutValue value)
     <*> pure (fromCardanoTxOutDatumHash datumHash)
 
-toCardanoTxOut :: C.NetworkId -> P.TxOut -> Either ToCardanoError (C.TxOut C.CtxTx C.AlonzoEra)
-toCardanoTxOut networkId (P.TxOut addr value datumHash) =
+toCardanoTxOut :: C.NetworkId -> Map P.DatumHash P.Datum -> P.TxOut -> Either ToCardanoError (C.TxOut C.CtxTx C.AlonzoEra)
+toCardanoTxOut networkId datums (P.TxOut addr value datumHash) =
     C.TxOut <$> toCardanoAddress networkId addr
             <*> toCardanoTxOutValue value
-            <*> toCardanoTxOutDatumHash datumHash
+            <*> cardanoDatumHash
+  where
+    cardanoDatumHash =
+      case flip Map.lookup datums =<< datumHash of
+        Just datum -> pure $ C.TxOutDatum C.ScriptDataInAlonzoEra (toScriptData $ P.getDatum datum)
+        Nothing    -> toCardanoTxOutDatumHash datumHash
+
+toScriptData :: P.BuiltinData -> C.ScriptData
```

Comment:
Isn't this the same as `toCardanoScriptData`?

Reply:
Indeed, I've fixed that